File: 2025/103. Binary Tree Zigzag Level Order Traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        zigzag(root, res, 0);
        return res;
    }

    private void zigzag(TreeNode root, List<List<Integer>> res, int depth) {
        if (root == null)
            return;
        if (res.size() <= depth) {
            res.add(new ArrayList<>());
        } 
        List<Integer> list = res.get(depth);
        if (depth % 2 == 0)
            list.add(root.val);
        else
            list.add(0, root.val);
        zigzag(root.left, res, depth + 1);
        zigzag(root.right, res, depth + 1);
    } 
}

File: 2025/103. Binary Tree Zigzag Level Order Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        res, queue = [], deque([root])
        odd_lvl = True
        while queue:
            level = deque()
            for _ in range(len(queue)):
                node = queue.popleft()
                if odd_lvl:
                    level.append(node.val)
                else:
                    level.appendleft(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            res.append(list(level))
            odd_lvl = not odd_lvl
        return res